src/sentry_core.c

```diff
@@ -785,4 +799,170 @@ sentry_transaction_finish(sentry_value_t tx)
     // This takes ownership of the transaction, generates an event ID, merges
     // scope
     return sentry__capture_event(tx);
+
+fail:
+    sentry_value_decref(tx);
+    return sentry_uuid_nil();
+}
+
+sentry_value_t
+sentry_set_span(sentry_value_t transaction)
+{
+    sentry_uuid_t nil_uuid = sentry_uuid_nil();
+    sentry_value_t sentinel = sentry__value_new_internal_uuid(&nil_uuid);
+    SENTRY_WITH_SCOPE_MUT (scope) {
+        sentry_value_decref(sentinel);
+        sentinel = sentry__value_clone(
+            sentry_value_get_by_key(transaction, "span_id"));
+        sentry_value_decref(scope->span);
+        scope->span = transaction;
+    }
+    return sentinel;
+}
+
+sentry_value_t
+sentry_span_start_child(
+    sentry_value_t transaction_or_span, char *operation, char *description)
+{
+    size_t max_spans = SENTRY_SPANS_MAX;
+    SENTRY_WITH_OPTIONS (options) {
+        max_spans = options->max_spans;
+    }
+
+    sentry_value_t parent;
+    // Parent is transaction on scope.
+    bool tx_on_scope = sentry_value_get_type(transaction_or_span)
+        == SENTRY_VALUE_TYPE_STRING;
+    if (tx_on_scope) {
+        // Changes to spans on a transaction requires no flushing.
+        SENTRY_WITH_SCOPE_MUT_NO_FLUSH (scope) {
+            // There isn't an active transaction. This span has nothing to
+            // attach to.
+            if (sentry_value_is_null(scope->span)) {
+                SENTRY_DEBUG(
+                    "no transaction available to create a child under");
+                return sentry_value_new_null();
+            }
+            parent = sentry__value_clone(scope->span);
+        }
+    } else {
+        parent = transaction_or_span;
+    }
+
+    if (!sentry_value_is_null(sentry_value_get_by_key(parent, "timestamp"))) {
+        SENTRY_DEBUG("span's parent is already finished, not creating span");
+        goto fail;
+    }
+
+    // Aggressively discard spans if a transaction is unsampled to avoid
+    // wasting memory
+    sentry_value_t sampled = sentry_value_get_by_key(parent, "sampled");
+    if (!sentry_value_is_true(sampled)) {
+        SENTRY_DEBUG("span's parent is unsampled, not creating span");
+        goto fail;
+    }
+    sentry_value_t spans = sentry_value_get_by_key(parent, "spans");
+    // This only checks that the number of _completed_ spans matches the number
+    // of max spans. This means that the number of in-flight spans can exceed
+    // the max number of spans.
+    if (sentry_value_get_length(spans) >= max_spans) {
+        SENTRY_DEBUG("reached maximum number of spans for transaction, not "
+                     "creating span");
+        goto fail;
+    }
```

Comment:
this is probably going to receive a lot of contention. i don't really see how it makes sense to let users continue to create and populate spans if they've already hit their span limit. this is already short circuiting if the parent span meets certain bad criteria so users are already expected to handle situations where a null is returned during span creation.

Reply:
IMO the idea should be that you should *always* be able to get the trace_id/span_id to attach to outgoing distributed tracing requests, even if you discard the span internally for whatever reason. So the span needs to "exist" internally.

Reply:
What doesn't line up for me is the fact that existing SDKs _do_ pre-emptively drop spans if they've reached the limit. At the same time, those SDKs also support distributed tracing which suggests to me that one of the two situations is happening:
- existing SDKs do not implement distributed tracing properly
- distributed traces do not continue directly off of spans

If it turns out that we're in the former scenario, the SDK TSC has acknowledged this as a known issue and we're making an active attempt to do better for this specific problem in the native SDK, I would be open to the idea of keeping the span internally in order to avoid breaking distributed tracing. I haven't found any documentation yet that suggests this is the case, however. It is also possible that I may have done a poor job of finding such documentation, so references/links to those would be welcome.